lib/sanity-util.js

```diff
@@ -91,128 +50,143 @@ function normalizeEntry({
-      return entry;
+    // If the value is null, we simply omit it from the formatted object.
+    if (normalizedValue === null) {
+      return result;
     }
 
-    visitedIds.add(entry._id);
+    // If the value is an array, we filter out any null members from it.
+    // If, by the end, we're left with an empty array, we omit it from
+    // the formatted object.
+    if (Array.isArray(normalizedValue)) {
+      const filteredValue = normalizedValue.filter(item => item !== null);
 
-    if (cache.has(entry._id)) {
-      return cache.get(entry._id);
+      if (filteredValue.length === 0) {
+        return result;
+      }
+
+      normalizedValue = filteredValue;
     }
 
-    const normalizedEntry = normalizeObject({
-      cache,
-      entry,
-      entriesById,
-      options,
-      visitedIds
-    });
+    return {
+      ...result,
+      [fieldName]: normalizedValue
+    };
+  }, initialObject);
 
-    cache.set(entry._id, normalizedEntry);
+  return result;
+}
 
-    return normalizedEntry;
+function normalizeValue({ cache, entriesById, options, value, visitedIds }) {
+  if (Array.isArray(value)) {
+    return value.map(childValue =>
+      normalizeValue({
+        cache,
+        entriesById,
+        options,
+        value: childValue,
+        visitedIds
+      })
+    );
   }
 
-  if (entry.toString() === "[object Object]") {
-    return normalizeObject({ cache, entry, entriesById, options, visitedIds });
+  if (typeof value !== "object" || value.toString() !== "[object Object]") {
+    return value;
   }
 
-  return entry;
-}
+  if (value._type === "slug") {
+    return value.current;
+  }
 
-function normalizeObject({ cache, entry, entriesById, options, visitedIds }) {
-  const initialObject = getMetadataBlock({ entry, options });
+  if (value._type === "reference") {
+    const { _ref: referenceId } = value;
+    const referencedEntry = entriesById[referenceId];
 
-  const result = Object.keys(entry).reduce((result, fieldName) => {
-    if (fieldName.startsWith("_")) {
-      return result;
-    }
+    if (!referencedEntry) return null;
 
-    const normalizedEntry = normalizeEntry({
+    return normalizeEntry({
       cache,
-      entry: entry[fieldName],
       entriesById,
+      entry: referencedEntry,
       options,
       visitedIds
     });
+  }
+
+  // At this point, `value` is an object, so we run it through the
+  // `normalizeObject` function.
+  const normalizedValue = normalizeObject({
+    cache,
+    entriesById,
+    object: value,
+    options,
+    visitedIds
+  });
+
+  if ((value._type === "file" || value._type === "image") && value.asset) {
+    const { asset, ...customFields } = normalizedValue;
 
     return {
-      ...result,
-      [fieldName]: normalizedEntry
+      ...asset,
+      ...customFields,
+
+      // We want to include a metadata block that represents the referenced
+      // entry, not the reference object.
+      ...getMetadataBlock({
+        entry: entriesById[value.asset._ref],
```

Comment:
@eduardoboucas, isn't `asset` already have the `__metadata` block?
The original `asset` is recursively resolved inside `normalizedObject` and then passed through `normalizeAsset({ entry, options })` here: https://github.com/stackbithq/sourcebit-source-sanity/blob/c8edd8db48ab705004fccabb83de4fcb13b2ee91/lib/sanity-util.js#L68-L70

So when the lines above where the asset is destructed `...asset`, it should already include its `__metadata`, no?

Reply:
or I am missing something? :)

Reply:
It does have a metadata block, but it’s generated from the reference object, not from the original/referenced asset. So we generate a new metadata block here, from the original entry, and override the other one.

Reply:
It does have a metadata block, but it’s generated from the reference object, not from the original/referenced asset. So we generate a new metadata block here, from the original entry, and override the other one.